kernel/arch/dreamcast/include/dc/g2bus.h

```diff
@@ -31,50 +32,184 @@
-   upon completion (in an interrupt context!). Returns <0 on error.
+#include <dc/fifo.h>
 
-   Known working combination :
+/** \defgroup g2b_channels         List of G2 Bus channels
+ 
+    AICA (SPU) is channel 0, BBA uses channel 1. CH2_DMA_G2CHN and 
+    CH3_DMA_G2CHN are not currently tied to any specific device.
 
-   g2chn = 0, sh4chn = 3 --> mode = 5 (but many other value seems OK ?)
-   g2chn = 1, sh4chn = 1 --> mode = 0 (or 4 better ?)
-   g2chn = 1, sh4chn = 0 --> mode = 3
+    \note 
+      A change in the implementation has rendered *_DMA_MODE and *_DMA_SHCHN 
+      obsolete.
 
-   It seems that g2chn is not important when choosing mode, so this mode parameter is probably
-   how we actually connect the sh4chn to the g2chn.
+      In the current implementation, *_DMA_MODE should always be set to zero 
+      (representing CPU_TRIGGER). There is also no involvement of SH4-DMA with
+      G2-DMA; therefore, the *_DMA_SHCHN values have been deprecated.
+      
+    @{
+*/
+#define SPU_DMA_G2CHN  0 /**< \brief AICA: G2 channel 0 */
+#define SPU_DMA_MODE   0 /**< \brief Unused; Here for compatibility */
+#define SPU_DMA_SHCHN  0 /**< \brief Unused; Here for compatibility */
 
-   Update : looks like there is a formula, mode = 3 + shchn
+#define BBA_DMA_G2CHN  1 /**< \brief BBA:  G2 channel 1 */
+#define BBA_DMA_MODE   0 /**< \brief Unused; Here for compatibility */
+#define BBA_DMA_SHCHN  0 /**< \brief Unused; Here for compatibility */
 
-*/
+#define CH2_DMA_G2CHN  2 /**< \brief CH2: G2 channel 2 */
+
+#define CH3_DMA_G2CHN  3 /**< \brief CH3: G2 channel 3 */
+/** @} */
 
-/* We use sh channel 3 here to avoid conflicts with the PVR. */
-#define SPU_DMA_MODE   6 /* should we use 6 instead, so that the formula is 3+shchn ?
-6 works too, so ... */
-#define SPU_DMA_G2CHN  0
-#define SPU_DMA_SHCHN  3
+/** \brief  G2Bus DMA direction
+
+    The direction you want the data to go. SH4 to AICA, BBA, etc use
+    SH4TOG2BUS, otherwise G2BUSTOSH4.
+*/
+#define SH4TOG2BUS   0
+#define G2BUSTOSH4   1
 
-/* For BBA : sh channel 1 (doesn't seem used) and g2 channel 1 to no conflict with SPU */
-#define BBA_DMA_MODE   4
-#define BBA_DMA_G2CHN  1
-#define BBA_DMA_SHCHN  1
+/** \brief  G2Bus DMA interrupt callback type.
 
-/* For BBA2 : sh channel 0 (doesn't seem used) and g2 channel 2 to no conflict with SPU */
-/* This is a second DMA channels used for the BBA, just for fun and see if we can initiate
-   two DMA transfers with the BBA concurently. */
-#define BBA_DMA2_MODE   3
-#define BBA_DMA2_G2CHN  2
-#define BBA_DMA2_SHCHN  0
+    Functions that act as callbacks when G2-DMA completes should be of this type.
+    These functions will be called inside an interrupt context, so don't try to
+    use anything that might stall.
 
+    \param  data            User data passed in to the g2_dma_transfer()
+                            function.
+*/
 typedef void (*g2_dma_callback_t)(ptr_t data);
```

Comment:
```suggestion
typedef void (*g2_dma_callback_t)(void* data);
```
Seriously, I had no idea wtf a `ptr_t` was just reading the API. Totally nonstandard. lol. It's very good that we're actually taking a generic userdata value here, but every sane API ever, including a bunch I've added to KOS take them as a `void*`, so I think we should be doing that as well to be consistent within KOS and idiomatic with the rest of C.

That being said, this is technically an API change, but I verified on compiler-explorer that the worst thing that can happen is they'll get warning if they're just passing an integer literal that it needs a cast.